pkg/streamingpromql/operators/vector_scalar_binary_operation.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/compat"
+	"github.com/grafana/mimir/pkg/streamingpromql/functions"
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// VectorScalarBinaryOperation represents a binary operation between an instant vector and a scalar such as "<expr> + 2" or "3 * <expr>".
+type VectorScalarBinaryOperation struct {
+	Scalar                   types.ScalarOperator
+	Vector                   types.InstantVectorOperator
+	ScalarIsLeftSide         bool
+	Op                       parser.ItemType
+	MemoryConsumptionTracker *limiting.MemoryConsumptionTracker
+
+	start     int64 // Milliseconds since Unix epoch
+	end       int64 // Milliseconds since Unix epoch
+	interval  int64 // In milliseconds
+	stepCount int
+
+	opFunc vectorScalarBinaryOperationFunc
+
+	expressionPosition posrange.PositionRange
+	emitAnnotation     functions.EmitAnnotationFunc
+	scalarData         types.ScalarData
+	vectorIterator     types.InstantVectorSeriesDataIterator
+}
+
+type vectorScalarBinaryOperationFunc func(scalar float64, vectorF float64, vectorH *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error)
+
+func NewVectorScalarBinaryOperation(
+	scalar types.ScalarOperator,
+	vector types.InstantVectorOperator,
+	scalarIsLeftSide bool,
+	op parser.ItemType,
+	start int64,
+	end int64,
+	interval int64,
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker,
+	annotations *annotations.Annotations,
+	expressionPosition posrange.PositionRange,
+) (*VectorScalarBinaryOperation, error) {
+	opFunc := arithmeticOperationFuncs[op]
+	if opFunc == nil {
+		return nil, compat.NewNotSupportedError(fmt.Sprintf("binary expression with '%s'", op))
+	}
+
+	b := &VectorScalarBinaryOperation{
+		Scalar:                   scalar,
+		Vector:                   vector,
+		ScalarIsLeftSide:         scalarIsLeftSide,
+		Op:                       op,
+		MemoryConsumptionTracker: memoryConsumptionTracker,
+
+		start:     start,
+		end:       end,
+		interval:  interval,
+		stepCount: stepCount(start, end, interval),
+
+		expressionPosition: expressionPosition,
+	}
+
+	b.emitAnnotation = func(generator functions.AnnotationGenerator) {
+		annotations.Add(generator("", expressionPosition))
+	}
+
+	if b.ScalarIsLeftSide {
+		b.opFunc = func(scalar float64, vectorF float64, vectorH *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+			return opFunc(scalar, vectorF, nil, vectorH)
+		}
+	} else {
+		b.opFunc = func(scalar float64, vectorF float64, vectorH *histogram.FloatHistogram) (float64, *histogram.FloatHistogram, bool, error) {
+			return opFunc(vectorF, scalar, vectorH, nil)
+		}
+	}
+
+	return b, nil
+}
+
+func (v *VectorScalarBinaryOperation) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	var err error
+	v.scalarData, err = v.Scalar.GetValues(ctx)
```

Comment:
Bit odd to do this here, but I understand why. I think this could warrant a comment that we take the opportunity to do this here instead of having to check if it's already been done in each NextSeries etc.

Reply:
I've added a comment in 138e66b